### PR TITLE
Client http server for token

### PR DESCRIPTION
**Pull request by lyda**
This passes back the token from the browser directly to the CLI. What will we all do with all those 5 seconds we save up?

There are a few lines marked `TODO`. A few are config options we could add, but really, how many knobs do we want the user to muck with? A few are possible security bits. Randomised paths, Randomised starting ports (and now that I think of it, it could just randomly pick ports the whole time instead of incrementally).

I also have a general Go question. The port check goroutine and the starting the http server goroutine both poke at the `listener.Srv.Addr` var (the former reads, the latter reads/writes). Is this allowed? Are there potential panics / race conditions this can cause? I can't really wrap these in a mutex since the `http.ListenAndServe` function is a blackbox - hence all the shenanigans to find the random port and communicate it back.

### CODE REVIEW

File: cmd/cashier/main.go [hunk 1]
```diff
@@ -49,22 +50,39 @@ func main() {
...
 	}
-	fmt.Printf("Your browser has been opened to visit %s\n", c.CA)
-	if err := browser.OpenURL(c.CA); err != nil {
+	authURL := c.CA
+	listener := &client.Listener{}
+	if c.AutoToken {
+		listener = client.StartHTTPServer()
+		if listener != nil {
+			authURL = fmt.Sprintf("%s?auto_token=%s",
+				c.CA, url.PathEscape(listener.ReceiverURL))
```
Comment (nsheridan) [inline cmd/cashier/main.go]: Does it make sense to allow a url here? What about just passing a port around instead of a full URL and assuming that the receiver will always be running on localhost? I fear that allowing a full URL could open this to some kind of abuse where a token can be sent somewhere else

File: server/auth/gitlab/gitlab.go [hunk 2]
```diff
@@ -1,13 +1,18 @@
 package gitlab
```
Comment (nsheridan) [inline server/auth/gitlab/gitlab.go]: This file looks unrelated to the change? Oh I see, this includes the gitlab changes too. Can you yank those from this PR easily? If not a rebase on this branch after merging the other PR might be the way to go

File: cmd/cashier/main.go [hunk 3]
```diff
@@ -49,22 +50,39 @@ func main() {
...
 	}
 
-	fmt.Print("Enter token: ")
-	scanner := bufio.NewScanner(os.Stdin)
-	var buffer bytes.Buffer
-	for scanner.Scan(); scanner.Text() != "."; scanner.Scan() {
-		buffer.WriteString(scanner.Text())
+	var token string
+	if listener != nil {
+		// TODO: Timeout?
```
Comment (nsheridan) [inline cmd/cashier/main.go]: I'm happy to let this run forever on someone's machine ¯\_(ツ)_/¯

**Comment by lyda** [comment]
> Is this an out of date comment? I can't see where this is being done.

Yes, sorry, that was removed.

File: server/auth/gitlab/gitlab.go (hunk 2 above)
Comment (lyda) [reply server/auth/gitlab/gitlab.go]: I needed it in order to test as the current gitlab support is broken. I can break it out, yes.

File: cmd/cashier/main.go (hunk 1 above)
Comment (lyda) [reply cmd/cashier/main.go]: Yep.

File: server/auth/gitlab/gitlab.go (hunk 2 above)
Comment (lyda) [reply server/auth/gitlab/gitlab.go]: Actually could you just merge the gitlab PR in and then I'll merge those changes in to this?

Comment (nsheridan) [reply server/auth/gitlab/gitlab.go]: Go for it

**Comment by lyda** [comment]
OK, changes from master moved in.

**Comment by lyda** [comment]
Finally tested this. It works until you start to:

* Turn on and off `auto_token`. I can see people doing this for different situations.
* Go from an old to a new cashier client.

So I think what I'm going to do in the `authed` function is see if the agent string is the cashier client from the http headers. If it is, then set the session value of the `auth_token` (which is `""` if there's no `auth_token`). If it's not then it's a redirect and we just go with whatever the redirect has to offer.

Not 100% sure on detecting the cashier client. Just looking for a Referer header might be enough.

Also, my original idea was to show the regular template page but have an `img src` tag in it that pointed to the client url. Then the client would return a success / fail image back. And the user would still have the ability to paste the token if there was an issue.

**Comment by nsheridan** [comment]
Implemented in https://github.com/nsheridan/cashier/commit/430471fa83442a8cbea2a3ecf6e2093fba03e063
